File: go/lc98.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	var check func(node *TreeNode, min, max int) bool
	check = func(node *TreeNode, min, max int) bool {
		if node == nil {
			return true
		}
		if node.Val <= min || node.Val >= max {
			return false
		}
		return check(node.Left, min, node.Val) &&
			check(node.Right, node.Val, max)
	}

	return check(root, math.MinInt, math.MaxInt)
}

// func isValidBST(root *TreeNode) bool {
// 	vals := make([]int, 0)
// 	var inorder func(node *TreeNode)
// 	inorder = func(node *TreeNode) {
// 		if node == nil {
// 			return
// 		}
// 		inorder(node.Left)
// 		vals = append(vals, node.Val)
// 		inorder(node.Right)
// 	}

// 	inorder(root)
// 	for i := 1; i < len(vals); i++ {
// 		if vals[i] <= vals[i-1] {
// 			return false
// 		}
// 	}
// 	return true
// }

File: java/com/lc/lc98.java
package lc;

import java.util.*;

public class lc98 {

    public boolean isValidBST(TreeNode root) {
        return check(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    public boolean check(TreeNode cur, long min, long max) {
        if (cur == null) {
            return true;
        }
        if (cur.val >= max || cur.val <= min) {
            return false;
        }
        return check(cur.left, min, cur.val)
                && check(cur.right, cur.val, max);
    }

    // List<Integer> tree = new ArrayList<>();

    // public boolean isValidBST(TreeNode root) {
    // inorder(root);
    // for (int i = 1; i < tree.size(); i++) {
    // if (tree.get(i) <= tree.get(i - 1)) {
    // return false;
    // }
    // }
    // return true;
    // }

    // public void inorder(TreeNode root) {
    // if (root == null) {
    // return;
    // }
    // inorder(root.left);
    // tree.add(root.val);
    // inorder(root.right);
    // }
}
